No more QgsGeometry pointers in the public API

m-kuhn opened the pull request:
Geometries are passed as const reference and returned by value.
This make using the API easier and reduces the risk of ownership
problems.

The overhead is minimal due to implicit sharing.

Fix https://github.com/qgis/qgis3.0_api/issues/68

## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

nyalldawson commented on src/core/qgspallabeling.cpp:
```diff
@@ -1429,22 +1429,20 @@ void QgsPalLayerSettings::registerFeature( QgsFeature &f, QgsRenderContext &cont
   if ( isObstacle )
   {
-    if ( obstacleGeometry && QgsPalLabeling::geometryRequiresPreparation( *obstacleGeometry, context, ct, doClip ? &extentGeom : nullptr ) )
+    if ( !obstacleGeometry.isNull() && QgsPalLabeling::geometryRequiresPreparation( obstacleGeometry, context, ct, doClip ? extentGeom : QgsGeometry() ) )
```
Fyi , there's a bool operator for QgsGeometry which does this too

m-kuhn replied:
Thanks, was wondering about that but thought "better safe than sorry" ;)

m-kuhn committed:
Fix QgsPoint(XY) sip bindings

m-kuhn commented:
Was there more info in the mail? can you forward?

3nids commented:
nope, you've got the whole reply...
my question is in the link

3nids commented:
I was just thinking, is there a way to compare a double to Py_None? i.e. what happens when you provide None to a double argument. You could check for None in some MethodCode and set the default value if it's a none.

m-kuhn commented:
You can make the param a `Py_Object*` (oor `SIP_PYTYPE`and then use `param == Py_None` to check and `Py_Float_As_Double` to convert to a double if applicable.
But I'm only 90% there yet.

3nids commented:
do you add a new signature with SIP_PYTYPE or do you just replace the double?
I had issue with trying to replace because there were still calls with doubles (maybe due to inherited classes calls)

m-kuhn committed:
Add an easy to use Z/M default value constructor to QgsPoint

m-kuhn commented on python/core/geometry/qgspoint.sip:
```diff
@@ -16,53 +16,42 @@ class QgsPoint: QgsAbstractGeometry
+//    static double NaN; // = std::numeric_limits<double>::quiet_NaN();
 
-    QgsPoint( double x = 0.0, double y = 0.0, double z = 0.0, double m = 0.0 );
+    QgsPoint( double x = 0.0, double y = 0.0, SIP_PYOBJECT z = Py_None, SIP_PYOBJECT m = Py_None ) [( double x = 0.0, double y = 0.0, double z = 0.0, double m = 0.0 )];
```
@3nids some success here :)

3nids replied:
well I would do it within an `#ifdef SIP_RUN` otherwise we could use something like:
`QgsPoint( double x = 0.0, double y = 0.0, double z = 0.0, double m = 0.0 ) SIP_SIGNATURE( double x = 0.0, double y = 0.0, SIP_PYOBJECT z = Py_None, SIP_PYOBJECT m = Py_None );`

m-kuhn committed:
Allow specifying a Docstring in #ifdef SIP_RUN blocks

m-kuhn committed:
Fix geometryvalidator

m-kuhn committed:
Fix sip bindings for QgsPoint

m-kuhn commented:
What do you think about the QgsPoint API with respect to Z/M constructors.

Currently (before this pull request), the following code looks unexpected to me.

```
QgsPoint(1, 2, 3).x() # returns 2
```

The reason is, that the first argument is treated as `QgsWkbTypes::Type`.

This pull request proposes to have an API like this:

```
QgsPoint(1,2).x() # returns 1
QgsPoint(1,2,3).z() # returns 3
QgsPoint(1,2, 3, 4).m() # returns 4
QgsPoint(1,2, m=4).x() # returns 4
```
I would propose to modify the previous API to read

```
QgsPoint(1,2,3,0,QgsWkbTypes.PointZ) # Creates a PointZ(1,2,3)
QgsPoint(1,2,3,type=QgsWkbTypes.PointZ)  # Creates a PointZ(1,2,3)
```

Does that sound reasonable?

m-kuhn commented:
Sorry, corrected

m-kuhn committed:
More intuitive QgsPoint python constructors

In python, the wkb type of a QgsPoint will by default be determined from
the provided parameters, where Z and M will be added as required if the
wkbType is Undefined.

    QgsPoint(x, y, z=nan, m=nan, wkbType=QgsWkbTypes.Undefined)

Thanks to the python API support of named parameters, it's also
straightforward to specify z, m and wkbType in any desired combination.

On the other hand, on C++ side it's often preferable to use

    QgsPoint(QgsWkbTypes::WkbType wkbType, double x, double y, double z, double m);

due to the lack of named parameters which make it harder to specify a
specific type and the advantage of typesafety that makes it possible to
verload the first constructor with this one.

m-kuhn committed:
Compare Z and M values in QgsPoint::operator== only when required

m-kuhn committed:
Fix point test

m-kuhn committed:
Projecting a 2D point with inclination assumes 0 for Z

m-kuhn commented:
QgsPoint improvements moved to #4712
